libvast/builtins/operators/export.cpp

```diff
@@ -0,0 +1,148 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/actors.hpp>
+#include <vast/atoms.hpp>
+#include <vast/catalog.hpp>
+#include <vast/concept/parseable/string/char_class.hpp>
+#include <vast/concept/parseable/vast/pipeline.hpp>
+#include <vast/error.hpp>
+#include <vast/logger.hpp>
+#include <vast/node_control.hpp>
+#include <vast/passive_partition.hpp>
+#include <vast/pipeline.hpp>
+#include <vast/plugin.hpp>
+#include <vast/query_context.hpp>
+#include <vast/table_slice.hpp>
+#include <vast/uuid.hpp>
+
+#include <arrow/type.h>
+#include <caf/attach_stream_source.hpp>
+#include <caf/scheduled_actor.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/timespan.hpp>
+#include <caf/typed_event_based_actor.hpp>
+
+namespace vast::plugins::export_ {
+
+class export_operator final : public crtp_operator<export_operator> {
+public:
+  auto operator()(operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    // TODO: Some of the the requests this operator makes are blocking, so we
+    // have to create a scoped actor here; once the operator API uses async we
+    // can offer a better mechanism here.
+    auto blocking_self = caf::scoped_actor(ctrl.self().system());
+    auto components
+      = get_node_components<catalog_actor, accountant_actor, filesystem_actor>(
+        blocking_self, ctrl.node());
+    auto [catalog, accountant, fs] = std::move(*components);
+    auto expr = trivially_true_expression();
+    if (auto pushdown = predicate_pushdown(expr)) {
+      expr = std::move(*pushdown).first;
+    } else {
+      VAST_VERBOSE("no expression found: launching 'export' to retrieve all "
+                   "persisted events");
+    }
```

Comment:
You could also make `to_string()` return `export | where <expr>`, but I think that this is quite hacky.

Reply:
I probably need to also add the expression to `make_operator`...

Reply:
We discussed this internally. The hackyness of `| where ...` will also go away with the upcoming serialization approach. Also, my mistake, I think you actually need 
~~~
    return std::pair{trivially_true_expression(),
                     std::make_unique<export_operator>(conjunction{expr_, expr})};
~~~
or something similar for full correctness. But otherwise 014416615d8e63b7a973d48b2c3c5e606735df1c looks good :+1: